README


This archive contains the distribution Net-Async-ControlChannel,
version 0.004:

  IO::Async implementation for simple key/value protocol

This software is copyright (c) 2013 by Tom Molesworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Net-Async-ControlChannel

0.004     2013-10-22 00:13:54 Europe/London
	No new features

	Bug fixes:
	* Delay connection notification until after the new client is set up,
	allowing the server to send welcome messages

0.003     2013-09-24 07:37:28 Europe/London
	No new features

	Test fixes:
	* Updated Dist::Zilla::Plugin::Test::Compile as described in RT88890,
	thanks to CHORNY for reporting

0.002     2013-09-20 05:55:04 Europe/London
	No new features

	Dependency fixes:
	* Include IO::Async in cpanfile

0.001     2013-09-19 06:55:14 Europe/London
	Initial CPAN release


lib/Net/Async/ControlChannel.pm
package Net::Async::ControlChannel;
# ABSTRACT: IO::Async implementation for simple key/value protocol
use strict;
use warnings;

our $VERSION = '0.004';

=head1 NAME

Net::Async::ControlChannel - L<IO::Async> support for ControlChannel protocol

=head1 VERSION

version 0.004

=head1 DESCRIPTION

Provides an L<IO::Async> implementation.

See documentation in:

=over 4

=item * L<Net::Async::ControlChannel::Server>

=item * L<Net::Async::ControlChannel::Client>

=item * L<Protocol::ControlChannel>

=back

=cut

use Net::Async::ControlChannel::Server;
use Net::Async::ControlChannel::Client;

1;

__END__

=head1 EXAMPLES

 #!/usr/bin/env perl
 use strict;
 use warnings;
 use IO::Async::Loop;
 use Net::Async::ControlChannel::Server;
 use Net::Async::ControlChannel::Client;
 use IO::Async::Timer::Periodic;
 
 my $loop = IO::Async::Loop->new;
 my $server = Net::Async::ControlChannel::Server->new(
 	loop => $loop,
 );
 $server->subscribe_to_event(
 	message => sub {
 		my $ev = shift;
 		my ($k, $v, $from) = @_;
 		warn "Server: Had $k => $v from $from\n";
 	},
 	connect => sub {
 		my $ev = shift;
 		my ($remote) = @_;
 		warn "Server: Client connects from $remote\n"
 	},
 	disconnect => sub {
 		my $ev = shift;
 		my ($remote) = @_;
 		warn "Server: Client disconnect from $remote\n"
 	}
 );
 {
 	$loop->add(my $timer = IO::Async::Timer::Periodic->new(
 		interval => 5,
 		on_tick => sub {
 			$server->dispatch('timer.tick' => time)
 		}
 	));
 	$timer->start;
 }
 my $f = $server->start->then(sub {
 	my $server = shift;
 	my $port = $server->port;
 	my $client = Net::Async::ControlChannel::Client->new(
 		loop => $loop,
 		host => $server->host,
 		port => $server->port,
 	);
 	$client->subscribe_to_event(
 		message => sub {
 			my $ev = shift;
 			my ($k, $v, $from) = @_;
 			warn "Client: Had $k => $v\n";
 			$client->dispatch('client.reply' => "$k:$v");
 		}
 	);
 	$client->start->on_done(sub {
 		my $client = shift;
 		$client->dispatch('client.ready' => time);
 	});
 });
 $loop->run;
 warn "finished\n";

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2013. Licensed under the same terms as Perl itself.


lib/Net/Async/ControlChannel/Client.pm
package Net::Async::ControlChannel::Client;
{
  $Net::Async::ControlChannel::Client::VERSION = '0.004';
}
use strict;
use warnings;
use parent qw(Mixin::Event::Dispatch);

=encoding utf8

=head1 NAME

Net::Async::ControlChannel::Client - L<IO::Async> support for L<Protocol::ControlChannel>.

=head1 VERSION

version 0.004

=head1 DESCRIPTION

Provides the client half for a control channel connection.

=cut

use Protocol::ControlChannel;
use Future;
use curry::weak;
use IO::Async::Stream;
use Scalar::Util qw(weaken);

=head1 METHODS

=cut

=head2 new

Instantiate a new client object.

Expects the following named parameters:

=over 4

=item * loop - the L<IO::Async::Loop> we will attach to

=item * host - which host we're connecting to

=item * port - the port to connect to

=back

Returns the instance.

=cut

sub new {
	my $class = shift;
	my %args = @_;
	my $self = bless {
		loop => (delete($args{loop}) or die 'no loop?'),
		%args
	}, $class;
	weaken($self->{loop});
	$self->{protocol} = Protocol::ControlChannel->new;
	$self;
}

=head2 loop

The L<IO::Async::Loop> object. Used internally.

=cut

sub loop { shift->{loop} }

=head2 proto

The L<Protocol::ControlChannel> instance. Used internally.

=cut

sub proto { shift->{protocol} }

=head2 connection

A L<Future> which resolves when the connection is established.

=cut

sub connection { shift->{connection} ||= Future->new }

=head2 start

Connects to the target host, returning a L<Future> which will
resolve once the connection is ready (this L<Future> is also
available via L</connection>).

=cut

sub start {
	my $self = shift;
	my $port = $self->{port};
	my $host = $self->{host};
	$self->loop->connect(
		addr => {
			ip => $host || 'localhost',
			family => 'inet',
			socktype => 'stream',
			port => $port || 0,
		},
	)->then(sub {
		my $sock = shift;
		my $stream = IO::Async::Stream->new(handle => $sock);
		$stream->configure(
			on_read => $self->curry::weak::incoming_message,
		);
		$self->loop->add($stream);
		$self->connection->done($stream);
	})->transform(done => sub { $self });
}

=head2 incoming_message

Called internally when we have data from the server.

=cut

sub incoming_message {
	my $self = shift;
	my (undef, $buffer, $eof) = @_;
	while(my $frame = $self->proto->extract_frame($buffer)) {
		$self->invoke_event(message => $frame->{key}, $frame->{value});
	}
	if($eof) {
		$self->invoke_event(disconnect => );
	}
	return 0;
}

=head2 dispatch

Dispatches the given key, value pair to the remote.

Expects two parameters:

=over 4

=item * $k - a Perl string representing the key we're sending over. Typically
this will be 'some.dotted.string'.

=item * $v - the value to send over, either a reference or a byte string.

=back

Unicode characters are allowed for the key, but if you want to send non-ASCII
text data in the content, it should be encoded explicitly:

 $cc->dispatch("utf₈.is.fine" => Encode::encode('UTF-8' => "ƃuıpoɔuǝ spǝǝu"));

=cut

sub dispatch {
	my $self = shift;
	my ($k, $v) = @_;
	my $holder;
	$holder = $self->connection->then(sub {
		my $stream = shift;
		my $f = Future->new;
		$stream->write(
			$self->proto->create_frame($k => $v),
			on_flush => sub { $f->done }
		);
		weaken $holder;
		$f;
	});
}

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2013. Licensed under the same terms as Perl itself.


lib/Net/Async/ControlChannel/Server.pm
package Net::Async::ControlChannel::Server;
{
  $Net::Async::ControlChannel::Server::VERSION = '0.004';
}
use strict;
use warnings;
use parent qw(Mixin::Event::Dispatch);

=encoding utf8

=head1 NAME

Net::Async::ControlChannel::Server - server implementation for L<Protocol::ControlChannel> using L<IO::Async>

=head1 VERSION

version 0.004

=head1 SYNOPSIS

 use Net::Async::ControlChannel::Server;
 use IO::Async::Loop;
 my $loop = IO::Async::Loop->new;
 my $server = Net::Async::ControlChannel::Server->new(
  loop => $loop,
 );
 $server->subscribe_to_event(
  message => sub {
   my $ev = shift;
   my ($k, $v, $from) = @_;
   warn "Server: Had $k => $v from $from\n";
  }
 );
 $server->start;
 $loop->run;

=head1 DESCRIPTION

Provides the server half for a control channel connection.

=cut

use Protocol::ControlChannel;
use Future;
use curry;
use curry::weak;
use List::UtilsBy qw(sort_by);

=head1 METHODS

=cut

use Scalar::Util qw(weaken refaddr);

=head2 new

Instantiates the server object. Will not establish the listener,
but does expect to receive an L<IO::Async::Loop> as the C<loop> named
parameter.

=cut

sub new {
	my $class = shift;
	my %args = @_;
	my $self = bless {
		loop => (delete($args{loop}) or die 'no loop?'),
		%args,
	}, $class;
	weaken($self->{loop});
	$self->{protocol} = Protocol::ControlChannel->new;
	$self;
}

=head1 ACCESSOR METHODS

=head2 loop

The L<IO::Async::Loop> instance we're (going to be) attached to.

=cut

sub loop { shift->{loop} }

=head2 host

Our host. Will be populated after L</start> has been called.

=cut

sub host { shift->{host} }

=head2 port

Our listening port. Will be populated after L</start> has been called.

=cut

sub port { shift->{port} }

=head2 proto

The L<Protocol::ControlChannel> instance. Mainly for internal use.

=cut

sub proto { shift->{protocol} }

=head2 clients

All currently-connected clients, as a list.

=cut

sub clients { sort_by { $_->{remote} } values %{ shift->{clients} } }

=head2 dispatch

Sends a message to all clients.

Expects two parameters:

=over 4

=item * $k - the key we're sending

=item * $v - the content (can be a ref, in which case it will be encoded
using whichever mechanism has been negotiated with the client)

=back

Returns a L<Future> which will resolve when we think we've delivered
to all connected clients.

=cut

sub dispatch {
	my $self = shift;
	return Future->new->done unless my @clients = $self->clients;

	my $void = !defined wantarray;
	my @pending;
	# Current implementation sends the same frame to all clients, so populate it once
	# TODO once encoding negotiation is implemented this will need to be per-client
	my $frame = $self->proto->create_frame(@_);
	for my $client (map $_->{stream}, $self->clients) {
		my @args;

		# No need to instantiate futures unless we're going to use the result somewhere
		unless($void) {
			my $f = Future->new;
			push @args, on_flush => sub { $f->done };
			push @pending, $f;
		}
		$client->write($frame, @args);
	}
	return if $void;

	Future->needs_all(@pending);
}

=head2 start

Start the listener. Will return a L<Future> which resolves with our
instance once the listening socket is active.

=cut

sub start {
	my $self = shift;
	$self->loop->listen(
		addr => {
			family   => 'inet',
			socktype => 'stream',
			ip       => $self->host || '0.0.0.0',
			port     => $self->port || 0,
		},
		on_listen => $self->curry::listen_active,
		on_stream => $self->curry::incoming_stream,
	);
	$self->listening;
}

=head2 listening

The L<Future> corresponding to the listening connection. Resolves with our
instance.

=cut

sub listening { shift->{listening} ||= Future->new }

=head2 listen_active

Called internally when the listen action is complete.

=cut

sub listen_active {
	my $self = shift;
	my $sock = shift;
	$self->{host} = $sock->sockhost;
	$self->{port} = $sock->sockport;
	$self->listening->done($self);
}

=head2 incoming_stream

Called internally for each incoming client.

=cut

sub incoming_stream {
	my $self = shift;
	my $stream = shift;
	my $remote = join ':', map $stream->read_handle->$_, qw(peerhost peerport);
	$self->{clients}{refaddr $stream} = {
		remote => $remote,
		stream => $stream,
	};
	$stream->configure(
		on_read => $self->curry::weak::incoming_message($remote),
	);
	$self->loop->add($stream);
	$self->invoke_event(connect => $remote);
}

=head2 incoming_message

Called internally when we have data from a client.

=cut

sub incoming_message {
	my ($self, $remote, $stream, $buffer, $eof) = @_;
	die "Message received from unknown client" unless exists $self->{clients}{refaddr $stream};
	while(my $frame = $self->proto->extract_frame($buffer)) {
		$self->invoke_event(message => $frame->{key}, $frame->{value}, $remote);
	}
	if($eof) {
		$self->invoke_event(disconnect => $remote);
		delete $self->{clients}{refaddr $stream};
	}
	return 0;
}

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2013. Licensed under the same terms as Perl itself.


Makefile.PL

use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.48;



my %WriteMakefileArgs = (
  "ABSTRACT" => "IO::Async implementation for simple key/value protocol",
  "AUTHOR" => "Tom Molesworth <cpan\@entitymodel.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.48"
  },
  "DISTNAME" => "Net-Async-ControlChannel",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Net::Async::ControlChannel",
  "PREREQ_PM" => {
    "Future" => "0.15",
    "IO::Async" => "0.50",
    "List::UtilsBy" => 0,
    "Mixin::Event::Dispatch" => "1.001",
    "Protocol::ControlChannel" => "0.001",
    "curry" => 0,
    "parent" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.004",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



